system-helper/flatpak-system-helper.c

```diff
@@ -141,6 +295,157 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
+  if (credentials == NULL)
+    return FALSE;
+
+  g_autoptr(GVariant) dict = g_variant_get_child_value (credentials, 0);
+  g_autoptr(GVariant) unixUserId = g_variant_lookup_value (dict, "UnixUserID", G_VARIANT_TYPE_UINT32);
+
+  if (unixUserId == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserId for the bus name: %s", sender);
+      return FALSE;
+    }
+
+  *uid_out = g_variant_get_uint32 (unixUserId);
+  return TRUE;
+}
+
+static gboolean
+handle_create_pull_cache_dir (FlatpakSystemHelper   *object,
+                              GDBusMethodInvocation *invocation,
+                              const gchar           *arg_ref,
+                              const gchar           *optional_commit,
+                              const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  FlatpakSystemHelperOngoingPull *new_pull;
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *user_folder = NULL;
+  g_autofree gchar *root_folder = NULL;
+  g_autofree gchar *flatpak_dir = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+  guint uid_pull_count = 0;
+  uid_t uid;
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (ongoing_pulls == NULL)
+    ongoing_pulls = g_ptr_array_new ();
+
+  if (!get_connection_uid (invocation, &uid, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      FlatpakSystemHelperOngoingPull *pull_tmp;
+      pull_tmp = g_ptr_array_index (ongoing_pulls, i);
+      if (pull_tmp->uid == uid)
+        {
+          uid_pull_count++;
+          // TODO: If there is a already a partial pull available, return that.
+        }
+    }
+
+  if (uid_pull_count > MAX_PULLS_PER_USER)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                             "Maximum number of pulls per user exceeded for UID: %u", uid);
+      return TRUE;
+    }
+
+  flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+  root_folder = g_mkdtemp_full (g_build_filename (flatpak_dir, "repo", "tmp", "flatpak-cache-XXXXXX", NULL), 0755);
+  user_folder = g_strdup_printf ("%s-bindfs", root_folder);
+  if (g_mkdir_with_parents (user_folder, 0755) == -1)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                             "Cannot create directory for bindfs mount at: %s", user_folder);
```

Comment:
Would be nice to include `g_strerror (errno)` in here. You may be interested in `glnx_throw_errno_prefix (&error, "Can't create directory for bindfs mount at %s", user_folder)`, here and the other places where `errno` is used. (I'd forgotten that Flatpak depends on libglnx which has these nice helper functions to turn `errno` into a GError.

Reply:
Addressed this in a172ca3f91aa5a003550a90de4dba0c6ab6b2441 with few others(which I have marked resolved)